src/test_helpers/xds.rs

```diff
@@ -0,0 +1,179 @@
+                        move |req: hyper::Request<hyper::Body>| srv.call(req),
+                    ))
+                }))
+                .await
+                .unwrap()
+        });
+        let xds_client = cfg.build(metrics, block_ready);
+        (tx, delta_tx, xds_client)
+    }
+}
+
+
+#[async_trait]
+impl AggregatedDiscoveryService for AdsServer {
+    type DeltaAggregatedResourcesStream = Pin<Box<dyn Stream<Item = Result<DeltaDiscoveryResponse, Status>> + Send>>;
+    type StreamAggregatedResourcesStream = Pin<Box<dyn Stream<Item = Result<DiscoveryResponse, Status>> + Send>>;
+    async fn stream_aggregated_resources(
+        &self,
+        request: tonic::Request<Streaming<DiscoveryRequest>>,
+    ) -> Result<
+        tonic::Response<Self::StreamAggregatedResourcesStream>,
+        tonic::Status,
+    > {
+        let current_response = match self.response.borrow().clone() {
+            Ok(resp) => {
+                resp
+            },
+            Err(e) => {
+                return Err(e);
+            }
+        };
+
+        info!("Received initial DiscoveryRequest");
+
+        let mut in_stream = request.into_inner();
+        let (tx, rx) = mpsc::channel(128);
+
+        tokio::spawn(async move {
+            while let Ok(result) = in_stream.message().await {
+                info!("processing request: {:?}", result);
+                match result {
+                    Some(_) => tx
+                        .send(Ok(current_response.clone()))
+                        .await
+                        .expect("working rx"),
+                    None => {
+                        warn!("test stream failed");
+                        break;
+                    }
+                }
+            }
+            info!("test stream ended");
+        });
+
+        let output_stream = ReceiverStream::new(rx);
+        Ok(Response::new(
+            Box::pin(output_stream) as Self::StreamAggregatedResourcesStream
+        ))        
+    }
+
+    
+    async fn delta_aggregated_resources(
+        &self,
+        request: tonic::Request<tonic::Streaming<DeltaDiscoveryRequest>>,
+    ) -> Result<
+        tonic::Response<Self::DeltaAggregatedResourcesStream>,
+        tonic::Status,
+    > {
+        let current_response = match self.delta_response.borrow().clone() {
+            Ok(resp) => {
+                resp
+            },
+            Err(e) => {
+                return Err(e);
+            }
+        };
+
+        info!("Received DeltaDiscoveryResponse request: {:?}", request);
+
+        let mut in_stream = request.into_inner();
+        let (tx, rx) = mpsc::channel(128);
+
+        tokio::spawn(async move {
+            while let Ok(result) = in_stream.message().await {
+                info!("processing result request: {:?}", result);
+                match result {
+                    Some(_) => tx
+                        .send(Ok(current_response.clone()))
```

Comment:
I think we are going to need tobe a little more sophisticated than a static "current_response" for this one. Might need to allow tests to pass a closure or something that emits a response based on the request

Reply:
Can you elaborate on your intuition here? The client should deterministically request resources based on the config, so how about a queue of responses for each case we want to test?  Then each test will essentially be a matter of defining the config for the client, and then writing the potential sequence of responses we want to make sure we handle. And finally validating the end state of our pulled resources on the client of course.

Reply:
Defining just a strict sequence may be tricky. I guess it can work, it just means our tests need to be really precise/coupled on their expectations of the client behavior rather than more "black box" testing where we just make sure the final state is correct.

We can start simple and expand as needed though